clarify lag tolerance use case (#9828)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/lag-tolerance.md

Source systems may update more frequently than downstream models need to rebuild. For example, a model used for daily reporting doesn't need to refresh more than once per day, even if new upstream data is available hourly.

> `lag_tolerance` defines how far upstream data can advance beyond the data already incorporated into a node before dbt State triggers a rebuild. This acts as a compute-saving buffer that helps you stay aligned with data freshness [Service Level Agreements (SLAs)](https://www.getdbt.com/blog/data-slas-best-practices) without unnecessary rebuilds. It supports two key scenarios:

Author: ```suggestion
`lag_tolerance` controls when dbt State rebuilds a node _after_ its upstream data changes. dbt compares the upstream data timestamp from the node's last build with the latest upstream data timestamp. If the difference is greater than the configured `lag_tolerance`, dbt rebuilds the node. Otherwise, it reuses the existing node. This acts as a compute-saving buffer that helps you stay aligned with data freshness [Service Level Agreements (SLAs)](https://www.getdbt.com/blog/data-slas-best-practices) without unnecessary rebuilds. It supports two key scenarios:
```

- **Aligning builds with SLA requirements**: `lag_tolerance` allows you to align model execution directly with data freshness SLA requirements, decoupling high-frequency upstream changes from downstream models that operate under wider, less demanding freshness requirements.
- **Protecting compute during upstream SLA breaches**: `lag_tolerance` protects your compute budget during freshness SLA breaches, preventing costly downstream rebuilds on static data when an upstream dependency fails its freshness SLA.

When dbt State evaluates whether to rebuild a node, it compares:

1. The latest upstream modification timestamp recorded when the node was last built.
2. The current latest upstream modification timestamp.
3. The difference between those two timestamps.

If the difference exceeds the configured `lag_tolerance`, dbt State rebuilds the node. Otherwise, it reuses the existing node rather than cloning or rebuilding it.

> This config accepts two value types:

Author: ```suggestion
The `lag_tolerance` config accepts two value types:
```

`lag_tolerance` controls when dbt State rebuilds a node _after_ its upstream data changes. dbt compares the upstream data timestamp from the node's last build with the latest upstream data timestamp. If the difference is greater than the configured `lag_tolerance`, dbt rebuilds the node. Otherwise, it reuses the existing node. This acts as a compute-saving buffer that helps you stay aligned with data freshness [Service Level Agreements (SLAs)](https://www.getdbt.com/blog/data-slas-best-practices) without unnecessary rebuilds. It supports two key scenarios:

- **Aligning builds with SLA requirements**: `lag_tolerance` allows you to align model execution directly with data freshness SLA requirements, decoupling high-frequency upstream changes from downstream models that operate under wider, less demanding freshness requirements.
- **Protecting compute during upstream SLA breaches**: `lag_tolerance` protects your compute budget during freshness SLA breaches, preventing costly downstream rebuilds on static data when an upstream dependency fails its freshness SLA.

> When dbt State decides whether to rebuild a node, it doesn't watch the clock — it watches how far the source data has moved on. It compares the source's freshness timestamp from the last time the node was actually built against the source's freshness timestamp right now. If that gap is bigger than your `lag_tolerance`, dbt rebuilds the node. If not, it reuses the node. See [How `lag_tolerance` is calculated](#how-lag_tolerance-is-calculated) for details.

Author: ```suggestion
When dbt State decides whether to rebuild a node, it doesn't watch the clock &mdash; instead it watches how far the source data has moved on. It compares the source's freshness timestamp from the last time the node was actually built against the source's freshness timestamp right now. If that gap is bigger than your `lag_tolerance`, dbt rebuilds the node. If not, it reuses the node. See [How `lag_tolerance` is calculated](#how-lag_tolerance-is-calculated) for details.
```

  lag_tolerance: "{{ '4h' if target.name == 'prod' else '7d' }}"
  ```

### How `lag_tolerance` is calculated

> dbt State compares two source data timestamps — not clock times:

Author: ```suggestion
dbt State compares two source data timestamps, not clock times:
```

<SimpleTable>

| Timestamp | What it is |
| --- | --- |

> | Source freshness at last build | How fresh the source was the last time dbt actually rebuilt this node |

Reviewer: This terminology took me awhile to absorb. Could be a me problem, but sharing an alternative that helped me think of this:

Previous Freshness Timestamp | Timestamp of when a dbt job run last ran the node due to a freshness trigger

Author: hey @savannah-dbt-labs , thanks so much! do you think using the point in time would be less abstract? eg.

```text
dbt State compares when the source had last been updated at two points:

| Point in time | Source timestamp used |
| --- | --- |
| Last node build | The source's latest update when dbt last rebuilt the node |
| Current run | The source's latest update detected during this run |

The best way to think about it is using the calculation:

lag = source update during current run − source update at last build
```

> | Source freshness now | How fresh the source is during this run |

Reviewer: This terminology took me awhile to absorb. Could be a me problem, but sharing an alternative that helped me think of this:

Current Freshness Timestamp | Timestamp of when a dbt job run most recently found the source was updated

</SimpleTable>

The best way to think about it is using the calculation:

```text
lag = source freshness now − source freshness at last build
```

If the lag exceeds `lag_tolerance`, dbt State rebuilds the node. Otherwise, it reuses the existing node.

Only a rebuild writes the baseline (source freshness at last build). A run that reuses the node leaves it untouched, so the next run subtracts against the same baseline:

```text
run → check lag (source now − baseline)

>    ├─ ≤ lag_tolerance → reuse   → baseline unchanged

Author: ```suggestion
   ├─ if lag is less than lag_tolerance → reuse   → baseline unchanged
```

>    └─ > lag_tolerance → rebuild → baseline = source now

Author: ```suggestion
   └─ if lag is greater than lag_tolerance → rebuild → baseline = source now
```

  lag_tolerance: "{{ '4h' if target.name == 'prod' else '7d' }}"
  ```

### How `lag_tolerance` is calculated
dbt State calculates lag by comparing two upstream data timestamps:

1. The upstream timestamp recorded when the node was last built.
2. The latest upstream timestamp available during the current invocation.

> The calculation is:

Author: ```suggestion
The best way to think about it is using the calculation:
```

```text
latest upstream timestamp - upstream timestamp recorded at the last build
```

If the difference exceeds `lag_tolerance`, dbt State rebuilds the node. Otherwise, it reuses the existing node.

> :::important

Author: ```suggestion
:::info `lag_tolerance measures timestamp differences 
```

> :::info `lag_tolerance measures timestamp differences 

Reviewer 2: ```suggestion
:::info `lag_tolerance` measures timestamp differences 
```

> `lag_tolerance` measures the difference between upstream data timestamps. It doesm't measure how long unprocessed data has been waiting or how much time has passed since the last dbt invocation.

Reviewer 2: ```suggestion
`lag_tolerance` measures the difference between upstream data timestamps. It doesn't measure how long unprocessed data has been waiting or how much time has passed since the last dbt invocation.
```

Author: ommgg why did i not use ai to check?! i had too much confidence in myself 😅

| Time | What happens | Lag calculation | dbt State result |
| --- | --- | --- | --- |
| `08:08` | The node is built. The source timestamp recorded for this build is `08:08`. | Starting point | Build |
| `08:33` | New data arrives in the source. | `08:33 - 08:08 = 25m` | No decision until the next dbt invocation |
| `09:08` | dbt State checks the source. Its latest timestamp is still `08:33`. | `08:33 - 08:08 = 25m` | Reuse: 25 minutes doesn't exceed `45m` |
| `10:08` | dbt State checks again. No additional source data has arrived. | `08:33 - 08:08 = 25m` | Reuse: the calculated lag is still 25 minutes |
| `10:30` | More data arrives in the source. | `10:30 - 08:08 = 2h 22m` | No decision until the next dbt invocation |
| `11:08` | dbt State checks the source. Its latest timestamp is now `10:30`. | `10:30 - 08:08 = 2h 22m` | Rebuild: the difference exceeds `45m` |

</SimpleTable>

Even though more than 45 minutes passed between the source update at `08:33` and the checks at `09:08` and `10:08`, the calculated lag remained 25 minutes. The node became eligible to rebuild only after the source timestamp advanced far enough beyond the `08:08` timestamp recorded during the last build.

> To rebuild a node whenever its upstream data changes, set:

Author: recommended but need to confirm if this is best practice

Author: @jeremyyeo

```yaml
state:
  lag_tolerance: 0s
```

> 

Author: ```suggestion
The following diagram shows how dbt State compares these timestamps to decide whether to reuse or rebuild a node:
```
